packages/telemetry/src/telemetry.ts

```diff
@@ -0,0 +1,210 @@
+import chalk from '@shuvi/utils/lib/chalk';
+import Conf from 'conf';
+import { BinaryLike, createHash, randomBytes } from 'crypto';
+
+import { getAnonymousMeta } from './helper/anonymous-meta';
+import { _postPayload } from './helper/post-payload';
+import { getRawProjectId } from './helper/project-id';
+
+// This is the key that specifies when the user was informed about anonymous
+// telemetry collection.
+const TELEMETRY_KEY_NOTIFY_DATE = 'telemetry.notifiedAt';
+
+// This is a quasi-persistent identifier used to dedupe recurring events. It's
+// generated from random data and completely anonymous.
+const TELEMETRY_KEY_ID = `telemetry.anonymousId`;
+
+// This is the cryptographic salt that is included within every hashed value.
+// This salt value is never sent to us, ensuring privacy and the one-way nature
+// of the hash (prevents dictionary lookups of pre-computed hashes).
+// See the `oneWayHash` function.
+const TELEMETRY_KEY_SALT = `telemetry.salt`;
+
+type TelemetryEvent = { eventName: string; payload: object };
+type EventContext = {
+  anonymousId?: string;
+  projectId?: string;
+  sessionId?: string;
+  [key: string]: unknown;
+};
+type EventMeta = { [key: string]: unknown };
+type EventBatchShape = {
+  eventName: string;
+  fields: object;
+};
+
+type RecordObject = {
+  isFulfilled: boolean;
+  isRejected: boolean;
+  value?: any;
+  reason?: any;
+};
+
+export class Telemetry {
+  private conf: Conf<any> | null;
+  private sessionId: string;
+  private rawProjectId: string;
+  private meta: EventMeta;
+  private context: EventContext;
+  private postEndpoint: string;
+
+  private queue: Set<Promise<RecordObject>>;
+
+  constructor({
+    version,
+    meta,
+    context,
+    postEndpoint
+  }: {
+    version?: string;
+    meta?: EventMeta;
+    context?: EventContext;
+    postEndpoint?: string;
+  }) {
+    try {
+      // `conf` incorrectly throws a permission error during initialization
+      // instead of waiting for first use. We need to handle it, otherwise the
+      // process may crash.
+      this.conf = new Conf({ projectName: 'shuvijs' });
+    } catch (_) {
+      this.conf = null;
+    }
+    this.sessionId = randomBytes(32).toString('hex');
+    this.rawProjectId = getRawProjectId();
+    this.postEndpoint = postEndpoint || 'TODO: update end point';
```

Comment:
I will suggest if the `postEndpoint` option is not provided, just skip the `_postPayload`.

Reply:
ok
https://github.com/shuvijs/shuvi/pull/473/commits/381a2a07b7c170dc41d49dda8bfc2fbee6a699c6